Asterix/optics/optical_systems.py

```diff
@@ -63,13 +63,64 @@ def __init__(self, modelconfig):
+
+                    if new_wl <= self.wavelength_0 - self.Delta_wav / 2 or new_wl >= self.wavelength_0 + self.Delta_wav / 2:
+                        raise ValueError(
+                            f"All [modelconfig]['mandatory_wls'] element must be in the range "
+                            "]wavelength_0 - Delta_wav / 2 , wavelength_0 + Delta_wav / 2[: "
+                            f"]{self.wavelength_0 - self.Delta_wav / 2},{self.wavelength_0 + self.Delta_wav / 2}[")
+                    # for each mandatory wavelength, we find the closest wavelength and replace it by the mandatory wavelength
+                    difference_array = np.absolute(self.wav_vec - new_wl)
+                    index_new_wave = difference_array.argmin()
+
+                    # we prevent the central wavelength to be replaced
+                    if self.wav_vec[index_new_wave] == self.wavelength_0 and new_wl < self.wavelength_0:
+                        index_new_wave -= 1
+                    if self.wav_vec[index_new_wave] == self.wavelength_0 and new_wl > self.wavelength_0:
+                        index_new_wave += 1
+
+                    self.wav_vec[index_new_wave] = new_wl
+
+            # we calculate the Riemann intervals to calculate the focal plane polychromatic intensity using the
+            # Riemann sum (https://en.wikipedia.org/wiki/riemann_sum) with the formula:
+            # I_poly = Sum(Imono[lambda_i] * riemannn_intervals_lengths[i]) / Delta_wav
+
+            # If there are no aditionnal "mandatory wavelengths" added, wavelengths are calculated so that the
+            # riemann intervals are all the same lengths (self.Delta_wav / self.nb_wav) with the simulation wavelengths
+            # centered on them. If we add one or more "mandatory wavelengths", the intervals are of different lengths
+            # and the simulation wavenlengths are not completely centered on them. It is preferable to use a high
+            # number of wavelengths in this case (nb_wav >= 7) so that the differences between each Riemann interval
+            # length are minimal.
+            riemann_interval_edges = np.zeros(self.nb_wav + 1)
+            riemann_interval_edges[0] = self.wavelength_0 - self.Delta_wav / 2
+            riemann_interval_edges[-1] = self.wavelength_0 + self.Delta_wav / 2
+
+            for i in range(self.nb_wav - 1):
+                riemann_interval_edges[i + 1] = (self.wav_vec[i] + self.wav_vec[i + 1]) / 2
+
+            self.riemannn_intervals_lengths = np.diff(riemann_interval_edges)
+
         else:
+            # If we are in the monochromatic case, we define self.wav_vec[0] = wavelength_0
+            # and riemannn_intervals_lengths[0] = 1 so we can apply the same formula.
             self.wav_vec = np.array([self.wavelength_0])
+            self.riemannn_intervals_lengths = np.array([self.Delta_wav])
             self.nb_wav = 1
 
         self.string_os = '_dimPP' + str(int(self.dim_overpad_pupil)) + "_resFP" + str(round(
```

Comment:
Comment by @rgalicher on lines 126+127 :
The science_sampling is the one for the central wavelength but when matrices are saved for "estimation_wl", the sampling value should change in the filename. It seems the value changes for the calculation. This is just a naming thing.

Reply:
ok done